Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -345,28 +437,120 @@ void CGContextSynchronize(CGContextRef context) {
 #pragma endregion
 
 #pragma region Global State - Transparency Layers
+HRESULT __CGContext::PushLayer(CGRect* rect) {
+    // Save our current D2D state block & drawing parameters.
+    // We will take a new GState for composing the layer later in PopLayer(): ...
+    RETURN_IF_FAILED(PushGState());
+
+    // ... this one.
+    auto& gStateForComposingLayer = CurrentGState();
+
+    ComPtr<ID2D1CommandList> commandList;
+    RETURN_IF_FAILED(deviceContext->CreateCommandList(&commandList));
+
+    deviceContext->SetTarget(commandList.Get());
+
+    // Copy the current layer's state to the new layer.
+    auto& oldLayer = _layerStack.top();
+
+    _layerStack.emplace(commandList.Get(), rect, oldLayer);
+    auto& newLayer = _layerStack.top(); // C++17 makes .emplace return a stack<T>::reference; alas, we are C++14.
+
+    // These properties are not to be preserved across transparency layers.
+    // https://developer.apple.com/library/content/documentation/GraphicsImaging/Conceptual/drawingwithquartz2d/dq_trans_layers/dq_trans_layers.html#//apple_ref/doc/uid/TP30001066-CH210-SW1
+    // Doing so would cause:
+    // * Double shadows
+    // * Double transparency
+    // * Double blending
+    // * Double clipping
+    auto& newDrawingState = newLayer.CurrentGState();
+    newDrawingState.alpha = 1.0;
+    newDrawingState.shadowColor = { 0.f, 0.f, 0.f, 0.f };
+    // newDrawingState.blendMode = kCGBlendModeNormal; // TODO GH#1389
+    newDrawingState.clippingGeometry = nullptr;
+    if (rect) {
+        CGRect transformedClippingRegion = CGContextConvertRectToDeviceSpace(this, *rect);
+
+        if (gStateForComposingLayer.HasShadow()) {
+            // The reference platform grows the layer's clipping region by the shadow's offset and its approximate width.
+            CGFloat xdelta = fabsf(gStateForComposingLayer.shadowOffset.width);
+            xdelta += gStateForComposingLayer.shadowBlur + 1;
```

Comment:
comment above, with auxiliary commentary in the pull request body.

---
In reply to: [90589260](https://github.com/Microsoft/WinObjC/pull/1496#discussion_r90589260) [](ancestors = 90589260)